From: Muqtadir <muqtxdir@gmail.com>
Date: Thu, 10 Nov 2022 15:24:13 +0530
Subject: info-overview: Replace os_logo with vanilla-logo-text-symbolic.

---
--- a/panels/info-overview/cc-info-overview-panel.c
+++ b/panels/info-overview/cc-info-overview-panel.c
@@ -67,7 +67,6 @@
   CcHostnameEntry *hostname_entry;
   CcListRow       *hostname_row;
   CcListRow       *memory_row;
-  GtkPicture      *os_logo;
   CcListRow       *os_name_row;
   CcListRow       *os_build_row;
   CcListRow       *os_type_row;
@@ -882,60 +881,6 @@
     open_software_update (self);
 }
 
-#if !defined(DISTRIBUTOR_LOGO) || defined(DARK_MODE_DISTRIBUTOR_LOGO)
-static gboolean
-use_dark_theme (CcInfoOverviewPanel *panel)
-{
-  AdwStyleManager *style_manager = adw_style_manager_get_default ();
-
-  return adw_style_manager_get_dark (style_manager);
-}
-#endif
-
-static void
-setup_os_logo (CcInfoOverviewPanel *panel)
-{
-#ifdef DISTRIBUTOR_LOGO
-#ifdef DARK_MODE_DISTRIBUTOR_LOGO
-  if (use_dark_theme (panel))
-    {
-      gtk_picture_set_filename (panel->os_logo, DARK_MODE_DISTRIBUTOR_LOGO);
-      return;
-    }
-#endif
-  gtk_picture_set_filename (panel->os_logo, DISTRIBUTOR_LOGO);
-  return;
-#else
-  GtkIconTheme *icon_theme;
-  g_autofree char *logo_name = g_get_os_info ("LOGO");
-  g_autoptr(GtkIconPaintable) icon_paintable = NULL;
-  g_autoptr(GPtrArray) array = NULL;
-  g_autoptr(GIcon) icon = NULL;
-  gboolean dark;
-
-  dark = use_dark_theme (panel);
-  if (logo_name == NULL)
-    logo_name = g_strdup ("gnome-logo");
-
-  array = g_ptr_array_new_with_free_func (g_free);
-  if (dark)
-    g_ptr_array_add (array, (gpointer) g_strdup_printf ("%s-text-dark", logo_name));
-  g_ptr_array_add (array, (gpointer) g_strdup_printf ("%s-text", logo_name));
-  if (dark)
-    g_ptr_array_add (array, (gpointer) g_strdup_printf ("%s-dark", logo_name));
-  g_ptr_array_add (array, (gpointer) g_strdup_printf ("%s", logo_name));
-
-  icon = g_themed_icon_new_from_names ((char **) array->pdata, array->len);
-  icon_theme = gtk_icon_theme_get_for_display (gdk_display_get_default ());
-  icon_paintable = gtk_icon_theme_lookup_by_gicon (icon_theme, icon,
-                                                   192,
-                                                   gtk_widget_get_scale_factor (GTK_WIDGET (panel)),
-                                                   gtk_widget_get_direction (GTK_WIDGET (panel)),
-                                                   0);
-  gtk_picture_set_paintable (panel->os_logo, GDK_PAINTABLE (icon_paintable));
-#endif
-}
-
 static void
 cc_info_overview_panel_class_init (CcInfoOverviewPanelClass *klass)
 {
@@ -952,7 +897,6 @@
   gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, hostname_entry);
   gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, hostname_row);
   gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, memory_row);
-  gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, os_logo);
   gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, os_name_row);
   gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, os_build_row);
   gtk_widget_class_bind_template_child (widget_class, CcInfoOverviewPanel, os_type_row);
@@ -983,8 +927,6 @@
 
   info_overview_panel_setup_overview (self);
   info_overview_panel_setup_virt (self);
-
-  setup_os_logo (self);
 }
 
 GtkWidget *
--- a/panels/info-overview/cc-info-overview-panel.ui
+++ b/panels/info-overview/cc-info-overview-panel.ui
@@ -12,9 +12,9 @@
                 <property name="maximum-size">192</property>
                 <property name="tightening-threshold">192</property>
                 <property name="child">
-                  <object class="GtkPicture" id="os_logo">
-                    <property name="can-shrink">False</property>
-                    <property name="alternative-text" translatable="yes">System Logo</property>
+                  <object class="GtkImage">
+                      <property name="icon-name">vanilla-logo-text-symbolic</property>
+                      <property name="pixel-size">256</property>
                   </object>
                 </property>
               </object>
